### PR TITLE
Add support to copy multiple csv files

### PR DESCRIPTION
**Pull request by acquamarin**
# Feature: copy multiple csv files into table
## 1. New grammar rule for copying files:
a. Copy from single file: `copy person from "person.csv"`
b. Copy from multiple files: `copy person from ["person1.csv", "person2.csv"]`
## 2. Note:
Copy from multiple parquet/arrow is still not supported.

### CODE REVIEW

File: src/include/storage/copy_arrow/copy_rel_arrow.h [hunk 1]
```diff
@@ -8,25 +8,26 @@ namespace kuzu {
 namespace storage {
 
 class CopyRelArrow : public CopyStructuresArrow {
+    enum class PopulateTaskType { populateAdjColumnsAndCountRelsInAdjListsTask, populateListsTask };
```
Comment (andyfengHKU) [inline src/include/storage/copy_arrow/copy_rel_arrow.h:11]: we always do 
```
enum class X : public uint8_t {
     A = 0,
     B = 1
}
```

**Review by ray6080 — APPROVED** [review]

File: src/include/storage/copy_arrow/copy_rel_arrow.h (hunk 1 above)
Comment (acquamarin) [reply src/include/storage/copy_arrow/copy_rel_arrow.h:11]: done

File: src/include/storage/copy_arrow/copy_structures_arrow.h [hunk 2]
```diff
@@ -23,35 +23,45 @@ namespace kuzu {
...
-    static void calculateListsMetadataAndAllocateInMemListPagesTask(uint64_t numNodes,
-        uint32_t elementSize, atomic_uint64_vec_t* listSizes,
-        ListHeadersBuilder* listHeadersBuilder, InMemLists* inMemList, bool hasNULLBytes,
-        const std::shared_ptr<spdlog::logger>& logger);
-
-    void countNumLines(const std::string& filePath);
-
-    arrow::Status countNumLinesCSV(std::string const& filePath);
+    inline FileBlockInfo getBlockInfoForFile(const std::string& fileName) {
+        return fileBlockInfos.at(fileName);
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/storage/copy_arrow/copy_structures_arrow.h:55]: done

File: src/storage/copy_arrow/copy_node_arrow.cpp [hunk 3]
```diff
@@ -12,77 +12,61 @@ namespace storage {
 CopyNodeArrow::CopyNodeArrow(CopyDescription& copyDescription, std::string outputDirectory,
     TaskScheduler& taskScheduler, Catalog& catalog, table_id_t tableID,
     NodesStatisticsAndDeletedIDs* nodesStatisticsAndDeletedIDs)
-    : CopyStructuresArrow{copyDescription, std::move(outputDirectory), taskScheduler, catalog},
+    : CopyStructuresArrow{copyDescription, std::move(outputDirectory), taskScheduler, catalog,
+          tableID},
       nodesStatisticsAndDeletedIDs{nodesStatisticsAndDeletedIDs} {
-    nodeTableSchema = catalog.getReadOnlyVersion()->getNodeTableSchema(tableID);
+    tableSchema = catalog.getReadOnlyVersion()->getTableSchema(tableID);
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/copy_arrow/copy_node_arrow.cpp:18]: done

File: src/storage/copy_arrow/copy_node_arrow.cpp [hunk 4]
```diff
@@ -104,76 +87,75 @@ arrow::Status CopyNodeArrow::populateColumns() {
...
-                break;
+    for (auto& filePath : copyDescription.filePaths) {
+        offset_t startOffset = getBlockInfoForFile(filePath).startOffset;
+        std::shared_ptr<arrow::csv::StreamingReader> csv_streaming_reader;
+        auto status = initCSVReaderAndCheckStatus(csv_streaming_reader, filePath);
+        std::shared_ptr<arrow::RecordBatch> currBatch;
+        int blockIdx = 0;
+        auto it = csv_streaming_reader->begin();
+        auto endIt = csv_streaming_reader->end();
+        while (it != endIt) {
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/copy_arrow/copy_node_arrow.cpp:98]: done

File: src/storage/copy_arrow/copy_structures_arrow.cpp [hunk 5]
```diff
@@ -10,131 +10,87 @@ namespace kuzu {
...
-            numPages++;
-        }
-        inMemList->getListsMetadataBuilder()->populateChunkPageList(chunkId, numPages,
-            inMemList->inMemFile->getNumPages() /* start idx of pages in .lists file */);
-        inMemList->inMemFile->addNewPages(numPages);
-    }
-    logger->trace("End: listsMetadata={0:p} listHeadersBuilder={1:p}",
-        (void*)inMemList->getListsMetadataBuilder(), (void*)listHeadersBuilder);
+uint64_t CopyStructuresArrow::writeColumnsToFilesAndUpdateStatistics() {
+    auto startTime = std::chrono::high_resolution_clock::now();
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/copy_arrow/copy_structures_arrow.cpp:45]: removed

File: src/storage/copy_arrow/copy_structures_arrow.cpp [hunk 6]
```diff
@@ -10,131 +10,87 @@ namespace kuzu {
...
-      copyDescription{copyDescription}, outputDirectory{std::move(outputDirectory)}, numBlocks{0},
-      taskScheduler{taskScheduler}, catalog{catalog}, numRows{0} {}
+      copyDescription{copyDescription}, outputDirectory{std::move(outputDirectory)},
+      taskScheduler{taskScheduler}, catalog{catalog}, numRows{0},
+      tableSchema{catalog.getReadOnlyVersion()->getTableSchema(tableID)} {}
+
+uint64_t CopyStructuresArrow::copy() {
+    logger->info(StringUtils::string_format("Copying {} file to table {}.",
+        CopyDescription::getFileTypeName(copyDescription.fileType), tableSchema->tableName));
+    auto startTime = std::chrono::high_resolution_clock::now();
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/copy_arrow/copy_structures_arrow.cpp:23]: removed

File: src/storage/copy_arrow/copy_structures_arrow.cpp [hunk 7]
```diff
@@ -10,131 +10,87 @@ namespace kuzu {
...
+    auto readTime = readFilesAndPopulateColumns();
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/copy_arrow/copy_structures_arrow.cpp:24]: renamed

File: src/storage/copy_arrow/copy_structures_arrow.cpp [hunk 8]
```diff
@@ -10,131 +10,87 @@ namespace kuzu {
...
-
-    case CopyDescription::FileType::PARQUET:
-        status = countNumLinesParquet(filePath);
-        break;
+    case CopyDescription::FileType::CSV: {
+        status = countNumLinesCSV(filePaths);
+    } break;
+    // Note: we only support copying a single arrow/parquet file to a table.
+    case CopyDescription::FileType::ARROW: {
+        status = countNumLinesArrow(filePaths[0]);
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/copy_arrow/copy_structures_arrow.cpp:60]: I think it is better to remove it after a discussion in the standup.

File: src/common/copier_config/copier_config.cpp [hunk 9]
```diff
@@ -1,49 +1,51 @@
...
 
 std::string CopyDescription::getFileTypeSuffix(FileType fileType) {
     return "." + getFileTypeName(fileType);
 }
 
-void CopyDescription::setFileType(std::string const& fileName) {
+void CopyDescription::setFileType() {
+    // We currently only support loading from files with the same type. Loading files with different
+    // types is not supported.
+    auto fileName = filePaths[0];
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/common/copier_config/copier_config.cpp:48]: moved the setFileType to binder

**Pull request merged**